katello-attach-subscription

```diff
@@ -99,70 +146,652 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
-  api = ApipieBindings::API.new({:uri => @options[:uri], :username => @options[:user], :password => @options[:pass], :api_version => '2', :timeout => @options[:timeout]}, {:verify_ssl => @options[:verify_ssl]})
+  # sanity check and fix of yaml file for subscription
+  # going to check if any id in the yaml do not exists, and expand the search keys
+  # sub will keep the originally requested subscriptions, sub_parsed the found one
+  if not cachepresent
+    puts "Subscription parsing started. Please be patient."
+    subsyamltotalentry = @yaml[:subs].count
 
-  req = api.resource(:home).call(:status)
-  if Gem::Version.new(req['version']) >= Gem::Version.new('1.11')
-    @host_resource = :hosts
-    @default_type = nil
-    @supports_registered_by = false
-  else
-    @host_resource = :systems
-    @default_type = 'Hypervisor'
-    @supports_registered_by = true
+    @yaml[:subs].each_with_index do |sub, subcurrentcount|
+      # search always for NORMAL sub as they could be attached booth hypervisor, physical and guest host
+      subfiltertype = "NORMAL"
+      puts " Started parsing entry #{subcurrentcount+1}/#{subsyamltotalentry}"
+      # check if the hash contains data. If not, skip it.
+      if not sub.is_a?(Hash)
+        puts "  YAML entry without valid data, like empty entry. No further actions on this entry."
+        next
+      end
+      # starting the process
+      if @options[:verbose]
+        puts " VERBOSE: started parsing for entry '#{sub["hostname"]}'"
+      end
+      if @options[:verbose]
+        puts " VERBOSE: subs look for host type '#{sub["type"]}', looking for '#{subfiltertype}' subscriptions"
+      end
+      # check only the sub section
+      # if sub is empty, auto-attach
+      if sub.has_key?('sub')
+        # initialize variables
+        sub_parsed = {}
+        desired_sub_hash = sub['sub']
+        if not desired_sub_hash.is_a?(Hash)
+          puts "  YAML entry without valid data, like empty subscription. As no valid subscriptions are listed, that will be deleted on the given hosts."
+          next
+        end
+        # for every product (hash key in yaml) check the desired subs
+        desired_sub_hash.each do |product, desidered_product_sub_array|
+          parsed_product_sub_array = []
+            puts "  Started parsing for product '#{product}'"
+          if @options[:debug]
+            puts "  DEBUG: in the desired_sub_hash, desidered_product_sub_array: #{desidered_product_sub_array}"
+          end
+          desidered_product_sub_array.each do |subscription_item|
+            parsed_subscription = []
+            if @options[:verbose]
+              puts "   VERBOSE: parsing subscription '#{subscription_item}'"
+            end
+            # fetch all the result searching subscription by their name, stored in subscription_item
+            # append 'name=' before the search criteria to search exactly that subscription name
+            search_options = 'name="'+subscription_item+'"'
```

Comment:
our YAML contains UUIDs, not names, does this search still work in that case? on Katello 3.7 I need to search for "id=<hash>" not "name=<hash>"